backupccl: return better error on failed backup

tbg opened the pull request:
The previous code would tend to return a context cancellation error when
a backup operation failed due to an error originating from an
ExportRequest, leaving no clue about the original failure condition
(this cost us a few hours today to discover when a user ran into it).

The root problem here is that `g.Err()` isn't supposed to exist and The
methods invoked by `Go` already have a `ctx` which they should use
instead, and outside of that you almost certainly want to check `Wait()`
to get a better error. Correspondingly, `g.Err()` has been removed to
avoid this pitfall in the future. This affected some other locations in
the codebase.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/32687)
<!-- Reviewable:end -->

mjibson commented on pkg/ccl/backupccl/restore.go:
```diff
@@ -810,8 +810,8 @@ func splitAndScatter(
 			}
 
 			select {
-			case <-g.Done:
```
Why is g.Done removed here but still used in backup.go?

tbg replied:
Replied in reviewable.

tbg reviewed (commented):
Fixed a pretty blatant bug that caused all uses of this new code to fail. I was always returning context canceled :see_no_evil: 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32687#-:-LSTfKnc20zKtTEVNnDk:b-nesp9d)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/ccl/backupccl/backup.go, line 678 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/32687#-LSTbsSj-hzd4I5l6LBz:-LSTbsSkAK8wwjChHB5A:b-wp5a4g) ([raw file](https://github.com/cockroachdb/cockroach/blob/5533e683467c23be8539090a02ee8c793a7d9d43/pkg/ccl/backupccl/backup.go#L678)):*
> ```Go
> 			select {
> 			case exportsSem <- struct{}{}:
> 			case <-g.Done:
> ```

@mjibson you're referring to this usage, right? The context passed to methods given to `GoCtx` is `g.ctx` and in particular will cancel when the group encounters an error. Here we only have the request context which doesn't know about the group cancelling (and would thus ship off more ExportRequests until the bitter end).


<!-- Sent from Reviewable.io -->

tbg committed:
backupccl: return better error on failed backup

The previous code would tend to return a context cancellation error when
a backup operation failed due to an error originating from an
ExportRequest, leaving no clue about the original failure condition
(this cost us a few hours today to discover when a user ran into it).

The root problem here is that `g.Err()` isn't supposed to exist and The
methods invoked by `Go` already have a `ctx` which they should use
instead, and outside of that you almost certainly want to check `Wait()`
to get a better error. Correspondingly, `g.Err()` has been removed to
avoid this pitfall in the future. This affected some other locations in
the codebase.

Release note: None

madelynnblue reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32687)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/ccl/backupccl/backup.go, line 678 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/32687#-LSTbsSj-hzd4I5l6LBz:-LSW-ByE7z8Oo3N0Em8Y:bf6e7u2) ([raw file](https://github.com/cockroachdb/cockroach/blob/5533e683467c23be8539090a02ee8c793a7d9d43/pkg/ccl/backupccl/backup.go#L678)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

@mjibson you're referring to this usage, right? The context passed to methods given to `GoCtx` is `g.ctx` and in particular will cancel when the group encounters an error. Here we only have the request context which doesn't know about the group cancelling (and would thus ship off more ExportRequests until the bitter end).
</blockquote></details>

My concern with the others is that calling `ctx.Done()` is much slower that `g.Done`. The entire point for g.Done to exist is for performance reasons, so I'm a bit worried about removing so many of them.


<!-- Sent from Reviewable.io -->

tbg reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32687)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/ccl/backupccl/backup.go, line 678 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/32687#-LSTbsSj-hzd4I5l6LBz:-LSZJNLy7IoBuwVlXfr9:bv95ki6) ([raw file](https://github.com/cockroachdb/cockroach/blob/5533e683467c23be8539090a02ee8c793a7d9d43/pkg/ccl/backupccl/backup.go#L678)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

My concern with the others is that calling `ctx.Done()` is much slower that `g.Done`. The entire point for g.Done to exist is for performance reasons, so I'm a bit worried about removing so many of them.
</blockquote></details>

Could you comment on the specific lines you think should have `g.Done` back? I just reviewed the diff but the other uses of `ctx.Done` are all inside of `GoCtx`. In `GoCtx`, you have the context which was wrapped by the errgroup, so it gets canceled when either the user-supplied context cancels, a goroutine returns an error, or Wait() returns, whichever happens first. This is the same as using `g.Done`. I might be missing something.


<!-- Sent from Reviewable.io -->

madelynnblue reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32687)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/ccl/backupccl/backup.go, line 678 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/32687#-LSTbsSj-hzd4I5l6LBz:-LSpSXyn0upJac5vf60Q:b387uh2) ([raw file](https://github.com/cockroachdb/cockroach/blob/5533e683467c23be8539090a02ee8c793a7d9d43/pkg/ccl/backupccl/backup.go#L678)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Could you comment on the specific lines you think should have `g.Done` back? I just reviewed the diff but the other uses of `ctx.Done` are all inside of `GoCtx`. In `GoCtx`, you have the context which was wrapped by the errgroup, so it gets canceled when either the user-supplied context cancels, a goroutine returns an error, or Wait() returns, whichever happens first. This is the same as using `g.Done`. I might be missing something.
</blockquote></details>

It's not a correctness problem, it's performance. Calling ctx.Done() on something in a tight loop (like dump.go where it's called on every row) is needlessly slow. g.Done (and g.Err()) were introduced to remove the need to call a function in a loop. The code in this PR might be correct but now g.Done is kinda a weird child without its g.Err sibling that it was meant to be used with. I just looked at the code a bit and I think I have a good way to solve this myself. LGTM on this PR; I'll fix up the other stuff myself so you are unblocked.


<!-- Sent from Reviewable.io -->

tbg reviewed (commented):
bors r=mjibson

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32687#-:-LSthkfU8YlL0sgkKgWD:b-yh73to)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/ccl/backupccl/backup.go, line 678 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/32687#-LSTbsSj-hzd4I5l6LBz:-LSthTWu9fTBqWzCaL2q:be17u80) ([raw file](https://github.com/cockroachdb/cockroach/blob/5533e683467c23be8539090a02ee8c793a7d9d43/pkg/ccl/backupccl/backup.go#L678)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

It's not a correctness problem, it's performance. Calling ctx.Done() on something in a tight loop (like dump.go where it's called on every row) is needlessly slow. g.Done (and g.Err()) were introduced to remove the need to call a function in a loop. The code in this PR might be correct but now g.Done is kinda a weird child without its g.Err sibling that it was meant to be used with. I just looked at the code a bit and I think I have a good way to solve this myself. LGTM on this PR; I'll fix up the other stuff myself so you are unblocked.
</blockquote></details>

Ack, thanks!


<!-- Sent from Reviewable.io -->
